### PR TITLE
Blacklisted accounts

### PR DESCRIPTION
**Pull request by vrrayz**
Fix for https://github.com/Joystream/pioneer/issues/4138

### CODE REVIEW

File: packages/ui/src/mocks/data/proposals.ts
Comment (thesan) [inline packages/ui/src/mocks/data/proposals.ts:245]: So with the `skip` argument, this is not necessary
```suggestion
```

File: packages/ui/src/mocks/helpers/asChainData.ts
Comment (thesan) [inline packages/ui/src/mocks/helpers/asChainData.ts:29]: Please remove the commented code on this file.

You're right the value can't be returned from the `asChainData`, instead something similar could be done on `asApiMethod` but it's not urgent right now...

**Review by thesan — CHANGES_REQUESTED** [review]

File: packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccounts.stories.tsx
Comment (vrrayz) [reply packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccounts.stories.tsx:30]: > Are you sure this is needed ?

This is the error that shows up if i don't include it 
image

I think its because the `useBalance` hook is called which requires the `all()` 
image

File: packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccounts.stories.tsx
Comment (vrrayz) [reply packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccounts.stories.tsx]: > I don't think this is needed:

After removing this and implementing other changes this error shows up
image

When writing the storybook test i came across  the error earlier on and i figured mocking the locks that way would be the fix for it which was i used this. In `toBalance`  the lockdetails are been used so i thought the storybook test needed such mock data
image

**Review by thesan — CHANGES_REQUESTED** [review]
Also it seems the balance encoding broke some tests. Please have a quick look. If there's no quick solution I'll fix it tomorrow.

Finally even if it works on storybook, this will require some actual QA before merging it. @ivanturlakov could you please create a bunch of accounts on Atlas dev transfer them some funds and Opt out of voting with them please ?
image

**Comment by ivanturlakov** [comment]
> ould you please create a bunch of accounts on Atlas dev transfer them some funds and Opt out of voting with them please ?

@thesan sure, what atlas instance should I use?

**Comment by thesan** [comment]
> @thesan sure, what atlas instance should I use?

@ivanturlakov I meant the Atlas dev network. However it seems that there is an opted out account already. So I guess nothing more is needed. Sorry I should have checked yesterday but I was convinced there was no opted out account on mainnet ATM.

@vrrayz please check: https://dao-git-fork-vrrayz-blacklisted-accounts-joystream.vercel.app/#/election/blacklisted-accounts there is a bug with the total counter, my guess is that it comes from: https://github.com/Joystream/pioneer/pull/4683#discussion_r1439576874 I'm surprised it didn't show on storybook (I guess the full app is triggering a lot more re-renders).

File: packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccount.tsx
Comment (thesan) [inline packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccount.tsx:17]: For most users `useBalance` will call `derive.balances.all`. RPC node queries are not cached and they should be reduced as much as possible. Since `useAddresses` is called on this page, this component shouldn't call `useBalance`. Instead:
```suggestion
export const BlacklistedAccount = ({ address, balance }: Prop) => {
```

File: packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccounts.tsx
Comment (thesan) [inline packages/ui/src/app/pages/Election/BlacklistedAccounts/BlacklistedAccounts.tsx:29]: As a result `useBalances` should be called on this component instead. Also I think the sum can be done here (no need for `TotalBalance`), and the paginated accounts should be `{ address: string, balance: BN | undefined }`:
```suggestion
  const balances = useBalances(addresses)
  const totalBalance = useMemo(
    () => balances.values().reduce((prev, value) => value?.total.add(prev) ?? prev, BN_ZERO),
    [balances]
  )
  const paginatedAccounts = useMemo(
    () =>
      votingOptOutAccounts
        .slice((page - 1) * ACCOUNTS_PER_PAGE, page * ACCOUNTS_PER_PAGE)
        .map((address) => ({ address, balance: balance.get(address)?.total }))
    [votingOptOutAccounts, page, balances])
  )
```

File: packages/ui/src/accounts/hooks/useBalance.ts
Comment (thesan) [inline packages/ui/src/accounts/hooks/useBalance.ts:31]: Small detail but please skip a line here:
```suggestion

export const useBalances = (addresses: Address[]): Map<string, Balances | undefined> => {
```

File: packages/ui/src/accounts/hooks/useBalance.ts
Comment (thesan) [inline packages/ui/src/accounts/hooks/useBalance.ts:38]: Even smaller detail (and I know it's something I wrote) but:
```suggestion
      if (definedBalances.has(address)) return
```

**Review by thesan — CHANGES_REQUESTED** [review]

**Review by thesan — APPROVED** [review]
LGTM!

I'm just waiting for #4713 to merge this

**Pull request merged**
